packages/notebook/src/model.ts

```diff
@@ -206,131 +249,172 @@ export class NotebookModel extends DocumentModel implements INotebookModel {
-    for (let key in metadata) {
+      // Update the metadata.
+      let metadata = { ...value.metadata };
       // orig_nbformat is not intended to be stored per spec.
-      if (key === 'orig_nbformat') {
-        continue;
+      delete metadata['orig_nbformat'];
+      let oldMetadata = { ...this.metadata };
+      for (let key in oldMetadata) {
+        oldMetadata[key] = null;
       }
-      this.metadata.set(key, metadata[key]);
-    }
-    this._ensureMetadata();
-    this.dirty = true;
+      let update = { ...oldMetadata, ...metadata };
+      DatastoreExt.updateField(
+        { ...this.data.record, field: 'metadata' },
+        update
+      );
+      this._ensureMetadata();
+    });
   }
 
   /**
    * Initialize the model with its current state.
    */
   initialize(): void {
-    super.initialize();
-    this.cells.clearUndo();
-  }
-
-  /**
-   * Handle a change in the cells list.
-   */
-  private _onCellsChanged(
-    list: IObservableList<ICellModel>,
-    change: IObservableList.IChangedArgs<ICellModel>
-  ): void {
-    switch (change.type) {
-      case 'add':
-        change.newValues.forEach(cell => {
-          cell.contentChanged.connect(this.triggerContentChange, this);
-        });
-        break;
-      case 'remove':
-        break;
-      case 'set':
-        change.newValues.forEach(cell => {
-          cell.contentChanged.connect(this.triggerContentChange, this);
-        });
-        break;
-      default:
-        break;
-    }
-    this.triggerContentChange();
+    /* No-op */
   }
 
   /**
    * Make sure we have the required metadata fields.
    */
   private _ensureMetadata(): void {
-    let metadata = this.metadata;
-    if (!metadata.has('language_info')) {
-      metadata.set('language_info', { name: '' });
+    DatastoreExt.withTransaction(this.data.record.datastore, () => {
+      let metadata = { ...this.metadata };
+      metadata['language_info'] = metadata['language_info'] || { name: '' };
+      metadata['kernelspec'] = metadata['kernelspec'] || {
+        name: '',
+        display_name: ''
+      };
+      DatastoreExt.updateField(
+        { ...this.data.record, field: 'metadata' },
+        metadata
+      );
+    });
+  }
+
+  private _onGenericChange(
+    sender: Datastore,
+    args: Datastore.IChangedArgs
+  ): void {
+    const change = args.change;
+    // Grab the changes for the schemas we are interested in.
+    const recordChange = change[this.data.record.schema.id];
+    const cellChange = change[this.data.cells.schema.id];
+    const outputChange = change[this.data.outputs.schema.id];
+    // If there was a change to any of the top-level items, emit a
+    // contentChanged signal.
+    if (recordChange) {
+      this._contentChanged.emit();
+      return;
     }
-    if (!metadata.has('kernelspec')) {
-      metadata.set('kernelspec', { name: '', display_name: '' });
+    // If there were any changes to the outputs, emit a contentChanged signal.
```

Comment:
The `contentChanged` signal is meant to tell the context whether to set the `isDirty` flag (among other things). The idea is that there are many view-related changes in the datastore which are not persisted to disk, and should not be considered content changes for the purposes of saving. This change handler is a bit tricky, but essentially filters the change, only emitting the signal for actual content changes, instead of things like cursor motion.